New rule: mixin and function named arguments

diego-codes opened the pull request:
Implements #181 

Two separate rules: `at-function-named-parameters` and `at-mixin-named-parameters` that implement the similar logic.

kristerkari commented:
Thanks! I'll try to get this reviewed asap :)

kristerkari commented on src/rules/at-function-named-arguments/README.md:
````diff
@@ -0,0 +1,115 @@
+
+```scss
+animation: animation($duration: 250ms) {
+//                 ↑
````
Small thing, but this arrow should point to the `$`.

kristerkari commented on src/rules/at-function-named-arguments/README.md:
````diff
@@ -0,0 +1,115 @@
+
+```scss
+.foo {
+  animation: animation($duration: 250ms);
````
Why is this considered a warning when the rule is set to `always-multiple-arguments`?

diego-codes replied:
That’s is a good point. Would `only-multiple-arguments` make sense?

kristerkari replied:
I did a search in `stylelint` repo, but I didn't see `only-` prefix used in any of the rules, so maybe we should also try to keep things consistent.

I think that we can keep the name as it is right now, or then do something like `"always"` with a secondary option `{ except: ["single-argument"] }`, but I think that that might just be more confusing.

I'll need to sleep the night first to be able to think clearly :)

diego-codes replied:
Hey, sorry I'm just getting back to this. I agree it's important to keep things consistent. 

If we add the secondary `{ "except": ["single-argument"] }` option, would using it along with `"never"` mean that single arguments have to be named?

Looking at `stylelint`'s guidelines another approach can be `{ "ignoreSingleArgument": true }` which basically would avoid checking single arguments for either `"never"` or `"always"`. I'm not sure that I like this either, but it's more clear then `"except"` in my opinion.

What do you think?

kristerkari replied:
Using `"ignore"` could work, so how about `{ "ignore": ["single-argument"] }`? That style would also be extendable in the future.

diego-codes replied:
That sounds good. I’ll make this change.

kristerkari commented on src/rules/at-mixin-named-arguments/README.md:
````diff
@@ -0,0 +1,115 @@
+
+```scss
+.foo {
+  @include animation($duration: 250ms);
````
Same question as with Sass functions: should this be a warning if the rule is set to `always-multiple-arguments´?

kristerkari commented:
Thanks @diego-codes! Sorry that it took me so long to review this, but I've been quite busy this week.

I added a couple of comments for things that I was unsure of. If you could check them out and after that I can merge this PR and release a new version.

Good job :+1:

diego-codes commented:
Hey @kristerkari, I removed the `"always-multiple-arguments"` primary option and added the `"ignore: ["single-argument"]"` secondary option based on our discussion. The number of tests basically doubled which may be excessive...

Let me know what you think!

kristerkari commented:
@diego-codes  Awesome! I would not be too worried about having a lot of tests. If you want to reduce duplication, you can use `.concat` like this rule does https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/double-slash-comment-empty-line-before/__tests__/index.js#L135-L148

Let me know if you still want to make any changes. If no changes are needed, I can merge this 👍

kristerkari commented on src/rules/at-function-named-arguments/index.js:
```diff
@@ -0,0 +1,133 @@
+      ruleName,
+      {
+        actual: expectation,
+        possible: ["always", "never", "always-multiple-arguments"]
```
` "always-multiple-arguments"` should be remove from here.

kristerkari commented on src/rules/at-mixin-named-arguments/index.js:
```diff
@@ -0,0 +1,112 @@
+      ruleName,
+      {
+        actual: expectation,
+        possible: ["always", "never", "always-multiple-arguments"]
```
` "always-multiple-arguments"` should be remove from here too.

kristerkari commented on src/rules/at-function-named-arguments/index.js:
```diff
@@ -0,0 +1,133 @@
+  "linear-gradient",
+  "radial-gradient",
+  "repeating-linear-gradient",
+  "repeating-radial-gradient"
```
Oh, btw. this function list seems incomplete, here's the full one:
https://github.com/EtaiG/css-functions-list

kristerkari replied:
and maybe you could also add a test that tests a CSS function that uses camelCase, for example `translateZ`.

diego-codes replied:
Good call!

diego-codes commented:
@kristerkari, I made the changes you requested. I think I'm okay with the tests as they are. I just added some to tests for CSS functions.

kristerkari commented on src/rules/at-function-named-arguments/index.js:
```diff
@@ -0,0 +1,125 @@
+export const messages = utils.ruleMessages(ruleName, {
+  expected: "Expected a named parameter to be used in function call",
+  rejected: "Unexpected a named parameter in function call",
+  rejectedSingle:
```
I had another look at the code and I noticed that `rejectedSingle` does not seem to be used anywhere. It's also in the other rule.

kristerkari commented on src/rules/at-function-named-arguments/__tests__/index.js:
```diff
@@ -0,0 +1,852 @@
+      `,
+      line: 3,
+      column: 9,
+      message: messages.unexpected,
```
Here and in the lines below the message should be `messages.rejected` as there is no `messages.unexpected` defined.

This also happens in the the tests for the other rule.

kristerkari commented:
Great work @diego-codes! I found still a couple of really small things, if you could fix those, then I can merge this and release a new version tomorrow. :)

diego-codes commented:
@kristerkari, sorry about that. That was really sloppy editing. Thanks for catching it. Let me know if you see anything else. Otherwise, I think it's good to go!
